vllm/core/scheduler.py

```diff
@@ -1583,6 +1583,43 @@ def _get_num_lookahead_slots(self, is_prefill: bool,
 
         return self.scheduler_config.num_lookahead_slots
 
+    def _get_budget_for_chunk(self, budget: SchedulingBudget) -> int:
+        """When doing chunked prefill, calculate the token budget for a single
+        chunk. This dynamically scales the chunk size down as the number of
+        sequences that require prefilling increases. This ensures that a single
+        sequence with a very large prompt to prefill doesn't take the entire
+        remaining token budget, allowing other sequences to prefill and decode
+        concurrently."""
+
+        # First get the number of sequences that require prefill
+        prefilling_seqs = 0
+        for i in range(len(self.running)):
+            for seq in self.running[i].seqs:
+                if seq.data.stage == SequenceStage.PREFILL:
+                    prefilling_seqs += 1
+        # Queued requests will also need to prefill
+        prefilling_seqs += len(self.waiting)
+
+        # Get the current remaining token budget
+        remaining_token_budget = budget.remaining_token_budget()
+
+        if prefilling_seqs == 0:
+            # Return immediately if there are no sequences that require prefill
+            return remaining_token_budget
+
+        # calculate a chunk size that shares it evenly across sequences that
+        # need to prefill
+        chunk_size = int(remaining_token_budget / prefilling_seqs)
+        # Ensure the chunk size is at least the minimum configured by the
+        # user, to limit the number of requests doing prefill
+        chunk_size = max(chunk_size, self.scheduler_config.min_chunk_size)
+        # And cap that at our actual budget so we don't spend tokens we
+        # don't have.
+        chunk_size = min(remaining_token_budget, chunk_size)
+
+        logger.debug("Prefill chunk size: %d", chunk_size)
```

Comment:
Do not use any logging. Even `logger.debug` would introduce latency overheads in the critical path.